knownvalue/not_null.go

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package knownvalue
+
+import (
+	"fmt"
+)
+
+var _ Check = notNull{}
+
+type notNull struct{}
+
+// CheckValue determines whether the passed value is nil.
+func (v notNull) CheckValue(other any) error {
+	if other == nil {
+		return fmt.Errorf("expected non-nil value for NotNull check, got: %T", other)
+	}
+
+	return nil
+}
+
+// String returns the string representation of notNull.
+func (v notNull) String() string {
+	return "not-null"
+}
+
+// NotNull returns a Check for asserting the value passed
+// to the CheckValue method is not nil.
```

Comment:
Curious q: Does this require any sort of mentioning SDKv2 behavior of storing `""` vs `nil` and how `knownvalue.NotNull` won't detect that?

I don't fully understand the old SDKv2 behavior so this question may be off-base

Reply:
I'm a little reluctant to mention any SDK/v2-specific behaviour within the `knownvalue` package itself. Of course if the consensus is that we should make an explicit statement here regarding SDK/v2 behaviour then I'm happy to do that, but I'd be interested in hearing a little more detail around exactly what should appear in the Go docs under these circumstances. In terms of the behaviour of `knownvalue.NotNull` then it should be the same irrespective of whether SDK/v2 or the framework have been used to write the provider, as the check is just looking to see whether the value is not nil/null in the state or plan. 

The built-in state checks for `TestCheckResourceAttrSet`, and `TestCheckNoResourceAttr` are different in that null values are removed from the state that is retrieved and used for evaluating `TestCheckFunc`(s). For example, the [random provider](https://github.com/hashicorp/terraform-provider-random/blob/main/internal/provider/resource_string_test.go#L808), uses, [TestCheckNoResourceAttr](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-testing@v1.6.0/helper/resource#TestCheckNoResourceAttr) to verify that an optional attribute (override_special) does not appear in state. This is a little peculiar, given that optional attributes that are not defined in configuration are stored in state as null. Investigation as to why this usage of [TestCheckNoResourceAttr](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-testing@v1.6.0/helper/resource#TestCheckNoResourceAttr) executes successfully reveals that immediately prior to the TestStep.Check being run, the state is retrieved by calling [getState()](https://github.com/hashicorp/terraform-plugin-testing/blob/main/helper/resource/testing_new.go#L462). This calls shimStateFromJson(), and ultimately [shimmedFlatmap.AddEntry()](https://github.com/hashicorp/terraform-plugin-testing/blob/main/helper/resource/state_shim.go#L290) which omits adding an entry for the attribute in the returned terraform.State if the attribute is nil, which is the case if the attribute has been omitted from the configuration, or explicitly set to null in the configuration.

Hope I'm not off the mark with this. Let me know if we need to discuss this further.

Reply:
Yeah that all makes sense to me. Probably best to avoid and if confusion arises, any documentation about `knownvalue` usage with SDKv2 would be best served in the SDKv2 section itself 😄